### PR TITLE
feat: modern-js switch on rspack 

### PR DESCRIPTION
**Pull request by GiveMe-A-Name**
## Description

feat: using rspack to compiler modern-js project

## Types of changes

- [ ] Docs change / Dependency upgrade
- [ ] Bug fix
- [x] New feature / Improvement
- [ ] Refactoring
- [ ] Breaking change

## Checklist

- [x] I have added changeset via `pnpm run change`.
- [ ] I have updated the documentation.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: packages/builder/builder-rspack-provider/src/types/rspack.ts [hunk 1]
```diff
@@ -1,6 +1,9 @@
-import type { Compiler, RspackOptions } from '@rspack/core';
+import type * as rspack from '@rspack/core';
 
-export { Compiler };
+type Compiler = rspack.Compiler;
+type RspackOptions = rspack.RspackOptions;
+
+export type { Compiler, rspack };
```
Comment (chenjiahan) [inline packages/builder/builder-rspack-provider/src/types/rspack.ts:6]: Makybe the `rspack` type should be `Rspack` (In PascalCase), the `webpack` type is a special case.

File: packages/builder/builder-shared/src/types/bundlerConfig.ts [hunk 2]
```diff
@@ -22,6 +22,45 @@ type SplitChunks = Configuration extends {
   : never;
 
 type WebpackOptimization = NonNullable<Configuration['optimization']>;
+type WebpackResolve = NonNullable<Configuration['resolve']>;
+type WebpackOutput = NonNullable<Configuration['output']>;
+
+// fork from the @rspack/core
```
Comment (chenjiahan) [inline packages/builder/builder-shared/src/types/bundlerConfig.ts:28]: Why fork this type

File: packages/cli/plugin-bff/src/cli.ts [hunk 3]
```diff
@@ -16,7 +16,7 @@ import { registerModernRuntimePath } from './helper';
 const DEFAULT_API_PREFIX = '/api';
 const TS_CONFIG_FILENAME = 'tsconfig.json';
 
-export default (): CliPlugin<AppTools> => ({
+export default (): CliPlugin<AppTools<'webpack'>> => ({
```
Comment (chenjiahan) [inline packages/cli/plugin-bff/src/cli.ts:19]: 'webpack' should be a default value, users don't need to explicitly declare it.

File: packages/solutions/app-tools/src/index.ts [hunk 4]
```diff
@@ -110,7 +110,16 @@ export const buildCommand = async (
   }
 };
 
-export default (): CliPlugin<AppTools> => ({
+interface AppToolsOptions {
+  /** Specify the use what kind of bundler to compiler, default: `webpack` */
+  bundler?: 'experiment-rspack' | 'webpack';
```
Comment (chenjiahan) [inline packages/solutions/app-tools/src/index.ts:115]: `experiment-rspack` -> `experimental-rspack`

File: packages/solutions/app-tools/package.json [hunk 5]
```diff
@@ -75,6 +75,7 @@
     "@modern-js/builder-plugin-node-polyfill": "workspace:*",
     "@modern-js/builder-shared": "workspace:*",
     "@modern-js/builder-webpack-provider": "workspace:*",
+    "@modern-js/builder-rspack-provider": "workspace:*",
```
Comment (chenjiahan) [inline packages/solutions/app-tools/package.json:78]: `@modern-js/builder-rspack-provider` should not be declared in the `dependencies` field.

File: packages/cli/plugin-bff/src/cli.ts (hunk 3 above)
Comment (GiveMe-A-Name) [reply packages/cli/plugin-bff/src/cli.ts:19]: The AppTools CliPlugin has 3 types temporarily: `AppTools<'webpack'>` 、 `AppTools<'rspack'>` 、 `AppTools<undefined>`. The default type is `AppTools<undefined>`, and the CliPlugin can use SharedTypes in defines. This is means the CliPlugin not only use in the base of webpack but alse use in rspack base

File: packages/builder/builder-shared/src/types/bundlerConfig.ts (hunk 2 above)
Comment (GiveMe-A-Name) [reply packages/builder/builder-shared/src/types/bundlerConfig.ts:28]: 1. We need the Rspack type in `bulider-shared` to define some config or type
2. the `@rspack/core`  bundler size is so big that we can't  integration into `builder-shared`

so I fork the type temporarily

File: packages/solutions/app-tools/package.json (hunk 5 above)
Comment (GiveMe-A-Name) [reply packages/solutions/app-tools/package.json]: done

File: packages/solutions/app-tools/src/index.ts (hunk 4 above)
Comment (GiveMe-A-Name) [reply packages/solutions/app-tools/src/index.ts]: done

File: packages/builder/builder-rspack-provider/src/types/rspack.ts (hunk 1 above)
Comment (GiveMe-A-Name) [reply packages/builder/builder-rspack-provider/src/types/rspack.ts]: done

File: packages/cli/plugin-bff/src/cli.ts (hunk 3 above)
Comment (chenjiahan) [reply packages/cli/plugin-bff/src/cli.ts:19]: It sounds like a breaking change for Plugin developers.

Comment (GiveMe-A-Name) [reply packages/cli/plugin-bff/src/cli.ts:19]: Or we can make the `CliPlugin<'webpack'>` default? After all, `Rspack` is in experimental status. 

But with the development of the `Rspack`, It's  necessary to distinguish between 'webpack' Plugin and 'rspack' Plugin. 
When user is not specify the CliPlugin type, is should be shared CliPlugin.

Comment (chenjiahan) [reply packages/cli/plugin-bff/src/cli.ts:19]: Yes, we should make the `CliPlugin<'webpack'>` as a default value.

Comment (GiveMe-A-Name) [reply packages/cli/plugin-bff/src/cli.ts]: done

**Review by chenjiahan — APPROVED** [review]

**Review by 9aoy — APPROVED** [review]

**Pull request merged**